### DISCUSSION TITLE
Is it possible to use `Annotated`-based mapping alongside `relationship(foreign_keys=)`?

### DISCUSSION THREAD

**Discussion by connebs** [sqlalchemy/sqlalchemy#8286] [Usage Questions] [answered]
I'm testing out the SQLAlchemy 2 beta and have a lot of models that reference the `User` model in my schema, so I made this `Annotated` shorthand:

```python
user_indexed_fk = Annotated[
    UUID, mapped_column(PG_UUID_TYPE, sa.ForeignKey("users.id"), nullable=False, index=True)
]
```

Of course frequently I want to use that ForeignKey reference alongside `relationships`, and sometimes a model will have multiple ForeignKey columns pointing to the user table, so I need to specify `foreign_keys` like so:

```python
class Card(ModelBase):
    creator_id: Mapped[user_indexed_fk]
    creator: Mapped[User] = relationship(
        "User",
        foreign_keys=[creator_id],
        backref=backref("cards", cascade="all, delete-orphan"),
    )
    editor_id: Mapped[user_indexed_fk]
    editor: Mapped[User] = relationship(
        "User",
        foreign_keys=[editor_id],
        backref=backref("cards", cascade="all, delete-orphan"),
    )
```

But of course in the above example I'm getting a `NameError` because `creator_id` / `editor_id` are not defined. I also tried using a string for `foreign_keys` (e.g. `foreign_keys=["creator_id"]` and `foreign_keys="creator_id"`) but neither of those seemed to work either. Is this pattern possible in SQLA2 or do I need to bite the bullet and just manually assign the column var to a `mapped_column` whenever I use `foreign_keys`?

**Comment by zzzeek** [comment, marked as answer]
hi !

 thanks for testing the new code / docs !   it's very helpful to have people looking around and seeing that things are working, and if you got as far as the Annotated i hope that things up to that point were going smoothly.

for what you're doing here you can use the normal "deferred" styles, the string approach as usual is in the context of mapped classes:

```
        foreign_keys="Card.creator_id",
```

and that will work fine.   better yet since it's typed is a lambda:

```
        foreign_keys=lambda: Card.creator_id,
```

at the moment that will produce a typing error that I am fixing now but it will work at runtime.

finally you can also put a placeholder mapped_column() and use that directly too:

```
    editor_id: Mapped[user_indexed_fk] = mapped_column()
    editor: Mapped[User] = relationship(
        "User",
        foreign_keys=editor_id,
        backref=backref("editors", cascade="all, delete-orphan"),
    )

```

it all works!  except for that typing glitch I will fix now.

**Reply by connebs** [reply]
Ahh, for some reason I didn't think to prefix the column name with the model name – makes sense. That lambda solution is pretty slick as well.

Thanks for getting back to me so quickly!
